ILLReflectometry update and validation

darominski opened the pull request:
**Description of work.**

This PR implements all changes needed for the successful deployment in production of the reflectometry package for ILL instruments D17 and FIGARO. These were partially supported as of 2018, but with changes to NeXus structure and metadata, and not-finished gravity correction issue for FIGARO, they never saw use in production during a reactor cycle. This PR addresses these last issues to allow full user adoption.

Tasks completed:
- [x] Update loader so it can load most recent D17 and FIGARO data
- [x] Address discrepancy between COSMOS and Mantid results for coherent summation type observed at D17
- [x] Address discrepancy in resolution calculation for D17
- [x] Add option to save .lam files
- [x] Add option to export .lam files from DrILL (enabled by @gui-co in PR #34446) 
- [x] Add exporting of the reduction summary to an output configuration file
- [x] Validate the reduction of D17 and FIGARO with responsible scientists
- [x] Include gravity correction for FIGARO, benchmark with COSMOS
- [x] Update documentation
- [x] Write technique document for reflectometry at ILL

**To test:**
1. New system tests ensure the results are consistent with COSMOS.
2. Code review
3. Documentation review

Fixes #33582 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

Documentation updated, including a new technique document for the ILL reflectometry, as well as release notes. 

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

darominski committed:
[REFLILL] Move D17 chopper parameters metadata names to IPF

The placement of the relevant metadata to obtain information regarding
chopper1 and chopper2 speed and phases has changed. The solution
places the names for those metadata in the D17 IPF.

darominski committed:
[REFLILL] Fix missing chopper-sample distance metadata

in D17 NeXus after cycle 211.

darominski committed:
[REFLILL] Fixed constParameter Cpp warnings

darominski committed:
[REFLILL] Fixed chopper metadata fetching in reduction

For D17 data after nexus format change.

darominski committed:
[REFLILL] Clearer source names for chopper metadata

darominski committed:
[REFLILL] Replaced Stitch1DMany with Stitch algorithm

and also removed redundant 'UseManualScaleFactors' which is not needed
by the improved algorithm. System tests required an update.

darominski committed:
[REFLILL] Updated link to sample horizontal offset for FIGARO

Data after cycle 193 seems to use underscores for the metadata entry
instead of camelcase.

darominski committed:
[REFLILL] Fixed naming of chopper metadata for FIGARO reduction

darominski committed:
[REFLILL] Unit tests for loading v3 nexus D17 and FIGARO

darominski committed:
[REFLILL] Logbook parameters for D17, FIGARO

darominski committed:
[REFLILL] Corrected comment

Unit conversion is handled by the loader, so at that point the distance is in metres.

darominski committed:
[REFLILL] Changed source of chopper-sample distance

darominski committed:
[REFLILL] Remove redundant init override

darominski committed:
[REFLILL] Report log now contains all input numors for db and rflb

darominski committed:
[REFLILL] Fixed coherent process low Q drop

This was caused by cutting direct beam wavelength range to the specified
values and then using such cut distribution to normalise the full wavelength
range of reflected beam, leading to mismatch and reflectivity drop-off at low Q.

darominski committed:
[REFLILL] Fixed monitor index for D17

darominski committed:
Merge branch '33582_illreflectometry_update_validation' of github.com:mantidproject/mantid into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] 'Unpack' list of runs if it contains lists

darominski committed:
[REFLILL] Fix getting DAN from FIGARO data

darominski committed:
[REFLILL] Perform consistency check before data loading

Consistency is checked for sample angle (SAN) and detector angle (DAN) with tolerance
of 0.1 degree, and for the same type of measurement, direct beam or reflected beam,
that would be summed during or shortly after loading.

darominski committed:
[REFLILL] Alternative chopper params when duration is short

Added options to choppers' read rotation_speed and phase in place of speed_average, and phase_average,
respectively in D17 measurements that are shorter than 30 s.

darominski committed:
[REFLILL] Parameters merging and logging in multifile merging

Chopper average speed and phase will be now averaged weighted by measurement time
in cases when multiple files at the same angular position are merged. Normalisation
to time is now done via duration instead of time, which had to be changed to a time log
array.

darominski committed:
Removed redundant issue: the property is not optional

darominski committed:
Append extension only if not already specified

darominski committed:
New ASCII reflectometry extension possible: lam

This extension is functionally identical with MFT type but contains an additional
column with wavelength value related to each momentum exchange.

darominski committed:
Unit test for lam extension handling

darominski committed:
[DRILL] Added handling lam file extension for Reflectometry

This commit also allows the extensions to be specified as a list,
and extracts the logic of creating drill task for exporting data
to its own method.

darominski committed:
Merge branch 'main' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Reduction params saved to a txt file

Revelant reduction parameters, defined in the IPF, are going to be stored in an external
text file. This commit contains the preliminary list of logs to be saved for D17.

darominski committed:
Specified unit for new theta property

darominski committed:
[DRILL] Docstring for createTask method and rename

darominski committed:
Correct unit: m_theta contains angle in rad

darominski committed:
[REFLILL] Added total TOF distance to summary report

darominski committed:
[REFLILL] Replaced sample-to-distance correction with loaded value

In each NeXus v3 file, there is already the distance from sample
detector, which takes into account the correction for the sample offset,
so there is no need to calculate this value each time from motor settings.

darominski committed:
[REFLILL] Stop correcting chopper to sample distance for future data

Currently, the field containing the distance from the chopper to sample
incorrectly does not take into account the sample offset, which is going
to be taken into account in the future data.

darominski committed:
Merge remote-tracking branch 'origin/main' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Fixed saving configuration in polarised analysis

darominski committed:
Merge branch 'main' into 33582_illreflectometry_update_validation

darominski committed:
Merge branch '33582_illreflectometry_update_validation' of github.com:mantidproject/mantid into 33582_illreflectometry_update_validation

darominski committed:
[DRILL] Call _createTasks method

darominski committed:
Removed redundant suppressions for ILL code

darominski committed:
[REFLILL] Removed redundant detectorAngle methods and variables

darominski committed:
[REFLILL] Removed confusing sign convention for offset angle

darominski committed:
Merge branch 'main' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Removed positive det order convention for FIGARO

Previously, D17 had negative sign convention attached to the offset angle,
while FIGARO used positive sign convention, which more careful analysis
found incorrect.

darominski committed:
[REFLILL] Simplified getting sample-detector distance

darominski committed:
[REFLILL] Fixed correction type for position correction for FIGARO,

using SpecularReflectionPositonCorrect algorithm.

darominski committed:
[REFLILL] Introduced additional angle correction for FIGARO,

based on a difference between angles spanned between direct beam position and instrument centre,
and reflected beam position and instrument centre. This was found necessary to fix the observed
discrepancy with COSMOS code.

darominski committed:
[REFLILL] Corrected sign convention and DAN bragg angle source,

predominantly for Figaro

darominski committed:
[REFLILL] Fix typo in test spectruminfo ids

darominski committed:
Merge branch 'main' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Made documentation order consistent

darominski committed:
[REFILL] Refl beam centre will be set to beam axis in DAN

angle option. The beam centre position needs to be rotated in any case
to the expected difference between the direct beam's DAN and reflected beams',
so the convenient reference point is the beam axis.

darominski committed:
[REFLILL] Hide redundant parameters in FIGARO instrumentviewer

darominski committed:
Merge remote-tracking branch 'origin/main' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Replace numpy operations with mantid algorithm

darominski committed:
[REFLILL] Changed extension of the parameter output

darominski committed:
[REFLILL] Added foreground refitting if there is more than one numor

The method is a re-implementation from the C++ code in LoadILLReflectometry.

darominski committed:
[REFLILL] Corrected formatting of if clause

darominski committed:
[REFLILL] Fixed getting pixel width and updated system test data

darominski committed:
[REFLILL] Replaced hard-coded chopper radius with IPF parameter

darominski committed:
[REFLILL] Direct beam ws is now rebinned to reflected beam

in ReflectometryILLSumForeground. Previously, the approach was the opposite, and the direct beam
was the source of binning to the reflected one.

darominski committed:
[REFLILL] Added wavelength part of gravity correction.

This is steered by CorrectGravity property added to the autoprocess and preprocess algorithms.
The gravity correction algorithm is a direct re-implementation from IDL in COSMOS.

darominski committed:
[REFLILL] Added angular gravity correction.

This is an adaptation of the COSMOS behaviour in the Mantid workflow.

darominski committed:
[REFLILL] Improved angle correction cleanup

darominski committed:
[REFLILL] Improved formatting

darominski committed:
[REFLILL] Added system test for gravity correction in Figaro

darominski committed:
[REFLILL] Replaced getName() with name()

darominski committed:
[REFLILL] Set configuration export parameters for D17, FIGARO

darominski committed:
[REFLILL] Added refdown property to figaro workspaces

darominski committed:
[REFLILL] Removed commented out code

darominski committed:
[REFLILL] DAN position fix of refdown samples at FIGARO

darominski committed:
[REFLILL] Removed slits, proposal, added BGS3 to the FIGARO logbook

darominski committed:
[REFLILL] Added accumulated logs to configuration output file

darominski committed:
[REFLILL] Updated docs and diagrams with gravity correction workflow

darominski committed:
[REFLILL] Added CorrectGravity property to DrILL general settings

darominski committed:
[REFLILL] ReflectometryILL_common reorganisation

according to PEP 8, 257, and 484, to improve naming conventions, docstrings, and remove redundant code.

darominski committed:
[REFLILL] ReflectometryILLPreprocess reorganisation

according to PEP 8, 257, and 484, to improve naming conventions, docstrings, and ordering of methods,
and imports.

darominski committed:
[REFLILL] Reorganization of ReflectometryILLAutoProcess

according to PEP 8, 257, and 484 by adding and reformatting docstrings,
enforcing snake_case naming scheme, alphabetical order of methods,
and explicit imports.

darominski committed:
[REFLILL] Preprocess docstrings: consistency in parameters->arguments

darominski committed:
[REFLILL] Reorganization of ReflectometryILLSumForeground

according to PEP 8, 257, and 484 by extending docstrings and type hints,
enforcing snake_case variable and method naming scheme, and reordering
methods alphabetically.

darominski committed:
[REFLILL] Reorganization of ReflectometryILLConvertToQ

to follow PEP 8, 257, and 484 by adding type hinting and extending docstrings,
enforcing snake_case name scheme for variable and method names, and reordering
methods alphabetically.

darominski committed:
[REFLILL] Reorganization of ReflectometryILLPolarizationCor

according to PEP 8, 257, and 484 by adding type hints, extending and adding
docstrings, enforcing snake_case naming scheme for variable and method names,
and re-ordering methods alphabetically.

darominski committed:
[REFLILL] Fixed missing import in the preprocess

darominski committed:
Merge remote-tracking branch 'origin/drill_export_extensions' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Adapted extension specificationfor exporting outputs

darominski committed:
Merge branch 'main' into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Fixes rebinning for polarized reflectometry.

If the workspace to rebin and the output are the same, while that workspace is a Workspace2D and the workspace to match is a group, there is a naming conflict.

darominski committed:
[REFLILL] Removed explicit init from system tests

darominski committed:
[REFLILL] Added cycle 213 D17 quartz system test.

darominski committed:
[REFLILL] Added technique doc for ILL reflectometry,

and updated other reflectometry ILL algorithm documentation for consistency.

darominski committed:
Merge branch 'main' into 33582_illreflectometry_update_validation

darominski committed:
Merge branch '33582_illreflectometry_update_validation' of github.com:mantidproject/mantid into 33582_illreflectometry_update_validation

darominski committed:
[REFLILL] Added release notes for new features

DannyHindson commented on docs/source/techniques/ReflectometryILL.rst:
```diff
@@ -0,0 +1,519 @@
+
+A very basic data reduction would include a direct and a reflected beam measurement of the sample at one angle position, follow the following steps:
+
+#. Preprocessing of direct and reflected beams in :ref:`ReflectometryILLConvertToQ <algm-ReflectometryILLConvertToQ>`
```
Is the algorithm name correct here? I wondered if it should be ReflectometryILLPreProcess looking at the workflow diagram

darominski replied:
Yes indeed, this is certainly an error.

DannyHindson commented on docs/source/techniques/ReflectometryILL.rst:
```diff
@@ -0,0 +1,519 @@
+Polarisation correction
+=======================
+
+This algorithm is an ILL specific wrapper around :ref:`PolarizationEffiencyCor <algm-PolarizationEfficiencyCor>`. It is typically run in reflectometry reduction workflow, namely the workflow in the :ref:`ReflectometryILLAutoProcess <algm-ReflectometryILLAutoProcess>` algorithm, between :ref:`ReflectometryILLSumForeground <algm-ReflectometryILLSumForeground>` and :ref:`ReflectometryILLConvertToQ < ...
```
What is "this algorithm" here? Is this ReflectometryILLPolarizationCor? Perhaps worth making this clear

DannyHindson commented on docs/source/techniques/ReflectometryILL.rst:
```diff
@@ -0,0 +1,519 @@
+
+Below is the relevant workflow diagram describing polarization correction workflow..
+
+.. diagram:: ReflectometryILLPolarizationCor-v1_wkflw.dot
```
This workflow looks like a duplicate of the one a few lines above?

DannyHindson reviewed (changes requested):
I had some time free today so thought I'd had a look over this PR that was in the unreviewed list. No problem if you want someone from the ILL team to review also.

Nice job tidying up all the variable naming adding type hints. These changes mean the number of functionality changes aren't perhaps as large as the number of line changes suggest. I've had a look through and come up with a few bits and pieces but looks pretty close to me

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLAutoProcess.py:
```diff
@@ -125,440 +151,29 @@ def summary(self):
+            rb = self.getProperty(PropertyNames.RB00).value
+        dimensionality = len(rb)
         if dimensionality != 0:
             for property_name in PropertyNames.PROPETIES_TO_SIZE_MATCH:
```
Well spotted!

darominski committed:
[REFLILL] Summation type name now a const, cleaned up naming

darominski committed:
[REFLILL] Release note for removal of UseManualScaleFactors property

from the ReflectometryILLAutoProcess algorithm as it was redundant.

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLAutoProcess.py:
```diff
@@ -582,300 +197,640 @@ def validateInputs(self):
+        direct_beam_input = self._compose_run_string(db_run)
+        self.preprocess_direct_beam(direct_beam_input, direct_beam_name, angle_index)
+        sum_type = self.get_value(PropertyNames.SUM_TYPE, angle_index)
+        sum_type = 'SumInLambda' if sum_type == PropertyNames.INCOHERENT else 'SumInQ'
```
Will add them as constants to the `ReflectometryILL_common.py`.

darominski commented on instrument/D17_Parameters.xml:
```diff
@@ -22,14 +22,15 @@
 		<!-- Default monitor spectrum number for monitor normalisation. -->
 		<parameter name="default-incident-monitor-spectrum" type="int" visible="false">
-			<value val="257" />
+			<value val="256" />
```
This was incorrectly filled from the beginning, so that's a correction of an error. It fixes the normalisation procedure, so perhaps a release note could be added.

darominski committed:
[REFLILL] Release note for correcting D17 monitor spectrum no

darominski committed:
[REFLILL] Moved ThetaCorrection property name to prop class

darominski committed:
[REFLILL] Log accumulation wrapped into a function

in ReflectometryILLAutoProcess, previously two separate calls were made for log
accumulation in built-in log handler and the other to append log to the list to be
later flushed to an output configuration file.

darominski committed:
[REFLILL] Corrections to algorithms mentioned in docs and diagrams

with proper reference to ReflectometryILLPreProcess and ReflectometryILLPolarizationCor,
and removal of doubling of workflow diagram for polarization correction.

darominski committed:
[REFLILL] Import all of ReflILL common not just SampleLogs

darominski committed:
[REFLILL] Added cleanup for theta correction

darominski committed:
[REFLILL] Removed redundant cleanup class from commons.

This class doesn't seem to be used anywhere and cleanup is handled by methods coming from ILL_utilities.

darominski committed:
[REFLILL] Added exception handling in case file doesn't exist

when trying to retrieve sample angle or detector angle.

darominski committed:
[REFLILL] Improved test coverage of common methods

darominski commented on Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryILL_commonTest.py:
```diff
@@ -21,12 +21,12 @@ def testD17SlitWidthLogEntry(self):
+        slit3_width = run.get(common.slit_size_log_entry(instr_name, 2))
+        common.slit_sizes(ws)
+        self.assertEqual(slit2_width.value, run.getProperty(common.SampleLogs.SLIT2WIDTH).value)
+        self.assertEqual(slit3_width.value, run.getProperty(common.SampleLogs.SLIT3WIDTH).value)
```
This is completely true, I haven't paid attention to it before. I have just now improved the test coverage of the `ReflectometryILL_common` package. It should be about 100% now.

DannyHindson reviewed (dismissed):
Thanks for all the changes! All looks good now

darominski committed:
[REFLILL] Changed type of log data from bool to int

due to issue in NeXus library, it is not possible to save a boolean sample log data,
thus it needed to be replaced by an int (1 for True, 0 for False).

darominski commented:
Thank you for the review. I am afraid I would need to ask for a re-approval due to incoming commit that fixes an issue I wanted to address in a separate PR, but in the end the scale of the intervention necessary was found to be too large. The issue refers to the use of boolean sample logs, which due to an issue in NeXus library cannot be saved properly and thus would be lost.
